### PR TITLE
ebs scheduling

### PR DESCRIPTION
**Pull request by StrongMonkey**

### CODE REVIEW

**Comment by StrongMonkey** [comment]
https://github.com/rancher/cattle/pull/2580 + https://github.com/rancher/storage/pull/41 + https://github.com/rancher/agent/pull/151 + https://github.com/rancher/scheduler/pull/43 should address the scheduling issue for rancher-ebs. So if volume is created through rancher-ebs, the availability zone will be persisted and scheduler will make the decision to schedule the instance to hosts which are in the same AZ. Since it is based on https://github.com/rancher/scheduler/pull/43, this scheduling feature only applies to service containers.

File: code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstancePreStart.java [hunk 1]
```diff
@@ -58,6 +62,12 @@ public void doWithLockNoResult() {
                     });
                 }
             }
+            Map<String, Object> driver_opts = DataAccessor.fieldMap(v, VolumeConstants.FIELD_VOLUME_DRIVER_OPTS);
+            if (driver_opts.containsKey("ec2_az")) {
```
Comment (cjellick) [inline code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstancePreStart.java]: Can you use the constants here?

File: code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstancePreStart.java [hunk 2]
```diff
@@ -58,6 +62,12 @@ public void doWithLockNoResult() {
                     });
                 }
             }
+            Map<String, Object> driver_opts = DataAccessor.fieldMap(v, VolumeConstants.FIELD_VOLUME_DRIVER_OPTS);
```
Comment (cjellick) [inline code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstancePreStart.java]: Using _ in variable names isnt how we roll in java.

Comment (StrongMonkey) [inline code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstancePreStart.java]: I am sort of lost in golang+java+python now...

File: code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstancePreStart.java [hunk 3]
```diff
@@ -58,6 +62,12 @@ public void doWithLockNoResult() {
...
+            if (driver_opts.containsKey(VolumeConstants.EC2_AZ)) {
+                String zone = driver_opts.get(VolumeConstants.EC2_AZ).toString();
+                labels.put(VolumeConstants.SCALE_GROUP_LABEL, String.format("key=%s;%s=1", VolumeConstants.EC2_AZ_HOST_LABEL_KEY, zone));
```
Comment (cjellick) [inline code/iaas/logic/src/main/java/io/cattle/platform/process/instance/InstancePreStart.java]: Why do we have to do scale_per_group scheduling here? Shouldn't a simple hard affinity constraint to hosts with the matching AZ label work?

**Comment by cjellick** [comment]
Needs an integration test

**Comment by StrongMonkey** [comment]
@cjellick integration test is added.

**Comment by cjellick** [comment]
LGTM

**Pull request merged**